src/auth_middleware.rs

```diff
@@ -0,0 +1,132 @@
+use hmac::Mac;
+use http_body_util::combinators::UnsyncBoxBody;
+use std::mem;
+use std::task::{self, Poll};
+
+use anyhow::Context;
+use bytes::Bytes;
+use futures_util::future::BoxFuture;
+use hmac::Hmac;
+use http_body_util::{BodyExt, Full};
+use sha2::Sha256;
+use tower::{Layer, Service};
+
+use crate::config::Config;
+
+type Hmac256 = Hmac<Sha256>;
+
+#[derive(Debug, Clone, Default)]
+pub struct AuthLayer {
+    pub shared_secret: Vec<String>,
+}
+
+// TODO do I need the layer?
+impl AuthLayer {
+    pub fn new(config: &Config) -> Self {
+        Self {
+            shared_secret: config.grpc_shared_secret.clone(),
+        }
+    }
+}
+
+impl<Inner> Layer<Inner> for AuthLayer {
+    type Service = AuthService<Inner>;
+
+    fn layer(&self, service: Inner) -> Self::Service {
+        AuthService::new(self.shared_secret.clone(), service)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct AuthService<Inner> {
+    secret: Vec<String>,
+    inner: Inner,
+}
+
+impl<Inner> AuthService<Inner> {
+    pub fn new(secret: Vec<String>, inner: Inner) -> Self {
+        Self { secret, inner }
+    }
+}
+
+type TonicBody = UnsyncBoxBody<Bytes, tonic::Status>;
+
+impl<Inner> Service<http::Request<TonicBody>> for AuthService<Inner>
+where
+    Inner: Service<http::Request<TonicBody>, Response = http::Response<TonicBody>>
+        + Clone
+        + Send
+        + 'static,
+    Inner::Future: Send,
+{
+    type Response = Inner::Response;
+    type Error = Inner::Error;
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, cx: &mut task::Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: http::Request<TonicBody>) -> Self::Future {
+        let secret = self.secret.clone();
+        let mut inner = self.inner.clone();
+        mem::swap(&mut inner, &mut self.inner);
+
+        Box::pin(async move {
+            let (parts, body) = req.into_parts();
+            let body_bytes = body.collect().await.unwrap().to_bytes();
```

Comment:
So easy!

Reply:
Yeah, getting the futures to be compatible and types to line up was the tricky part. 😵